knora/dsplib/utils/xml_upload.py

```diff
@@ -843,65 +769,147 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
     if not incremental:
         resources, stashed_xml_texts, stashed_resptr_props = remove_circular_references(resources, verbose)
     else:
         stashed_xml_texts = dict()
         stashed_resptr_props = dict()
 
-    sipi_server = Sipi(sipi, con.get_token())
+    id2iri_mapping: dict[str, str] = {}
+    failed_uploads: list[str] = []
 
-    # get the project information and project ontology from the server
-    project = ResourceInstanceFactory(con, shortcode)
+    try:
+        id2iri_mapping, failed_uploads = upload_resources(verbose, resources, imgdir, sipi_server, permissions_lookup,
+                                                          resclass_name_2_type, id2iri_mapping, con, failed_uploads)
+    except BaseException as err:
+        handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+        exit(1)
+
+    # update the resources with the stashed XML texts
+    nonapplied_xml_texts = {}
+    if len(stashed_xml_texts) > 0:
+        try:
+            nonapplied_xml_texts = update_stashed_xml_texts(verbose, id2iri_mapping, con, stashed_xml_texts)
+        except BaseException as err:
+            handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+            exit(1)
+
+    # update the resources with the stashed resptrs
+    nonapplied_resptr_props = {}
+    if len(stashed_resptr_props) > 0:
+        try:
+            nonapplied_resptr_props = update_stashed_resptr_props(verbose, id2iri_mapping, con, stashed_resptr_props)
+        except BaseException as err:
+            handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+            exit(1)
+
+    # write log files
+    exit_code = 0
+    timestamp_str = datetime.now().strftime("%Y%m%d-%H%M%S")
+    write_id2iri_mapping(input_file, id2iri_mapping, timestamp_str)
+    if len(nonapplied_xml_texts) > 0:
+        write_stashed_xml_texts(nonapplied_xml_texts, timestamp_str)
+        exit_code = 1
+    if len(nonapplied_resptr_props) > 0:
+        write_stashed_respr_props(nonapplied_resptr_props, timestamp_str)
+        exit_code = 1
 
-    # create a dictionary to look up permissions
-    permissions_lookup: dict[str, Permissions] = {}
-    for key, perm in permissions.items():
-        permissions_lookup[key] = perm.get_permission_instance()
+    if failed_uploads:
+        print(f"Could not upload the following resources: {failed_uploads}")
+        exit_code = 1
 
-    # create a dictionary to look up resource classes
-    resclass_name_2_type: dict[str, type] = {}
-    for res_class_name in project.get_resclass_names():
-        resclass_name_2_type[res_class_name] = project.get_resclass_type(res_class_name)
+    exit(exit_code)
 
-    res_iri_lookup: dict[str, str] = {}
 
-    failed_uploads = []
+def try_sipi_upload(sipi_server: Sipi, filepath: str) -> dict[Any, Any]:
```

Comment:
this should typically be protected/private, I would say

Reply:
to be concrete: should I rename it to `def _try_sipi_upload()`?

Reply:
I would do it, but it's a matter of taste, so definitely no requirement for getting this merged :)